[Jobs] Add `huggingface-cli jobs` commands

lhoestq opened the pull request.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_3211). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

lhoestq commented on setup.py:
```diff
@@ -20,6 +20,7 @@ def get_version() -> str:
     "requests",
     "tqdm>=4.42.1",
     "typing-extensions>=3.7.4.3",  # to be able to import TypeAlias
+    "dotenv",
```
I hope this is fine as a new dependency - I'd have loved to not rely on anything but env variables parsing seems non trivial unfortunately

davanstrien replied:
There are no dependencies for the package (unless you install with a [cli] extras flag).  Source: https://libraries.io/pypi/python-dotenv + https://github.com/theskumar/python-dotenv/blob/16f2bdad2ebbaae72790514cce713d2d22ab0f7c/setup.py#L41

julien-c commented on src/huggingface_hub/commands/jobs/uv.py:
```diff
@@ -0,0 +1,223 @@
+                    print(f"Using existing repository: {repo_id}")
+            except RepositoryNotFoundError:
+                print(f"Creating repository: {repo_id}")
+                create_repo(repo_id, repo_type="dataset", exist_ok=True)
```
```suggestion
                create_repo(repo_id, repo_type="dataset", private=True, exist_ok=True)
```
to not risk leaking stuff...

julien-c commented on src/huggingface_hub/commands/jobs/uv.py:
It's fine as an experiment, but not a huge fan of the local file uploading to a remote repo..

Is there any way to either:
- pass the file content as an argument (string) to `uv` (and thus to the Jobs creation API)
- ask the infra team to add a new feature to the Jobs creation API where you can a dict of file name to file contents and they are exposed to the docker command? (not sure if it's feasible @christophe-rannou)

davanstrien replied:
> pass the file content as an argument (string) to uv (and thus to the Jobs creation API)
I don't think this is directly possible in UV at the moment. 

> ask the infra team to add a new feature to the Jobs creation API where you can a dict of file name to file contents and they are exposed to the docker command? (not sure if it's feasible @christophe-rannou)

Think this would be nice if it was possible. @christophe-rannou, would this be difficult to implement?

lhoestq commented:
> As I high-level comment, it'd be good to have all the API logic added to HfApi (and therefore callable from a Python script) and the CLI logic (e.g. args, result formatting, etc.) kept in the current ./src/huggingface_hub/commands/jobs folder. @lhoestq Let us know if you have bandwidth to work on this or if you want some help

I can take care of this for tomorrow

hanouticelina commented on src/huggingface_hub/commands/jobs/cancel.py:
```diff
@@ -0,0 +1,32 @@
+    def run(self) -> None:
+        username = whoami(self.token)["name"]
+        headers = build_hf_headers(token=self.token)
+        requests.post(
+            f"https://huggingface.co/api/jobs/{username}/{self.job_id}/cancel",
+            headers=headers,
+        ).raise_for_status()
```
I’d recommend using [get_session()](https://github.com/huggingface/huggingface_hub/blob/37d32780de87b4ebd3c0fdf99695cbae0d58daf7/src/huggingface_hub/utils/_http.py#L163) instead of instantiating raw `requests.Session()`. it ensures one session per thread which makes it more safe and it's also used consistently in the rest of the codebase.

hanouticelina reviewed (commented):
Very nice UX, I like it! 🔥 i left some initial comments.
As mentioned by @Wauplin, let's centralize the logic into `HfApi`:
```python
HfApi.run_job(...)
HfApi.list_jobs(...)
HfApi.inspect_job(...)
HfApi.cancel_job(...)
HfApi.fetch_job_logs(...)
```
that way, the CLI subcommands become lightweight wrappers and maybe we can put all the sub parsers (`run`, `ps`, `inspect`, `logs`, `cancel`, `uv`) inside one `src/huggingface_hub/commands/jobs.py` file.

julien-c commented:
@davanstrien @lhoestq the infra team is working on the "mount this file in the container" API feature btw

For the current prototype feature is there a way to tell uv to pass a bearer token when downloading the remote file? or maybe it's overkill?

I'm also ok with the big red warning that their file is going to be uploaded publicly.

davanstrien commented:
@julien-c 

>For the current prototype feature is there a way to tell uv to pass a bearer token when downloading the remote file? or maybe it's overkill?

It's not currently possible to do this directly in UV, but we hope they will add this as a general feature, allowing `uv run URL` to work out of the box with private HF URLs. 

Perhaps for a V1, we go with making the repository public and adding a warning?

Wauplin commented on src/huggingface_hub/commands/jobs.py:
```diff
@@ -0,0 +1,471 @@
+    def register_subcommand(parser: _SubParsersAction) -> None:
+        run_parser = parser.add_parser("run", help="Run a Job")
+        run_parser.add_argument("image", type=str, help="The Docker image to use.")
+        run_parser.add_argument("-e", "--env", action="append", help="Set environment variables.")
+        run_parser.add_argument("-s", "--secret", action="append", help="Set secret environment variables.")
+        run_parser.add_argument("--env-file", type=str, help="Read in a file of environment variables.")
+        run_parser.add_argument("--secret-env-file", type=str, help="Read in a file of secret environment variables.")
```
```suggestion
        run_parser.add_argument("-e", "--env", action="append", help="Set environment variables.")
        run_parser.add_argument("-s", "--secrets", action="append", help="Set secret environment variables.")
        run_parser.add_argument("--env-file", type=str, help="Read in a file of environment variables.")
        run_parser.add_argument("--secrets-file", type=str, help="Read in a file of secret environment variables.")
```

(nit) I'd rename to `--secrets` and `--secrets-file`

Wauplin commented on src/huggingface_hub/commands/jobs.py:
```diff
@@ -0,0 +1,471 @@
+        self.image: str = args.image
+        self.command: List[str] = args.command
+        self.env: dict[str, Optional[str]] = {}
+        for env_value in args.env or []:
+            self.env.update(dotenv_values(stream=io.StringIO(env_value)))
+        if args.env_file:
+            self.env.update(dotenv_values(args.env_file))
```
I feel that "env" should take precedence over "env_file" i.e. if someone hardcode a variable in its command line, it should not be overwritten by the file in case of conflict. Same for secrets.

(also possible to add a check to make `env` and `env_file` mutually exclusive)

Wauplin commented on src/huggingface_hub/hf_api.py:
```diff
@@ -9940,6 +9944,488 @@ def auth_check(
+                # Simple hash for uniqueness
+                script_hash = hashlib.md5(Path(script).read_bytes()).hexdigest()[:8]
+
+                return f"{username}/huggingface-cli-jobs-uv-run-{timestamp}-{script_hash}"
```
```suggestion
                return f"{username}/hf-jobs-uv-run-{timestamp}-{script_hash}"
```

(let's make it shorter)

Wauplin reviewed (commented):
Made a first pass on the PR, thanks for having moved everything in `HfApi` :)

lhoestq commented:
I fixed `huggingface-cli jobs uv run local_script.py` :)
It uploads to a private repository, and the job first downloads the file in python before running the `uv run` command.

I simplified the logic a bit to always upload to the same repository since it's a bit of a pain to remove the repos afterwards.

Taking a look at your comments now @Wauplin , thanks for the review !

hanouticelina reviewed (commented):
Thanks @lhoestq for the iteration! i left some minor comments

Wauplin commented on src/huggingface_hub/utils/_dotenv.py:
```diff
@@ -0,0 +1,50 @@
+import re
```
```suggestion
# AI-generated module (ChatGPT) 
import re
```

Wauplin commented on tests/test_utils_dotenv.py:
```diff
@@ -0,0 +1,63 @@
+from huggingface_hub.utils._dotenv import load_dotenv
```
```suggestion
# AI-generated module (ChatGPT)
from huggingface_hub.utils._dotenv import load_dotenv
```

Wauplin commented:
Note that we're working on switching from `huggingface-cli ...` to `hf ...` :) (long awaited feature). See https://github.com/huggingface/huggingface_hub/pull/3229. Nothing to do in this PR, I will make sure to adapt in a future PR once both are merged. We will release the `hf` CLI and the Jobs API at the same time.

lhoestq commented:
I took your comments into account, let me know if anything is missing :)

I may add a dedicated docs page showcasing the HfApi features for jobs. But feel free to merge already if you'd like and I can do it in another PR

Wauplin commented on src/huggingface_hub/hf_api.py:
```diff
@@ -9940,6 +9943,458 @@ def auth_check(
+
+    def list_jobs(
+        self,
+        timeout: Optional[int] = None,
```
```suggestion
        *,
        timeout: Optional[int] = None,
```

Wauplin commented on src/huggingface_hub/hf_api.py:
```diff
@@ -9940,6 +9943,458 @@ def auth_check(
+
+    def inspect_job(
+        self,
+        job_id: str,
```
```suggestion
        *,
        job_id: str,
```

Wauplin commented on src/huggingface_hub/hf_api.py:
```diff
@@ -9940,6 +9943,458 @@ def auth_check(
+        return JobInfo(**response.json(), endpoint=self.endpoint)
+
+    def cancel_job(
+        self,
```
```suggestion
        self,
        *,
```

Wauplin commented on src/huggingface_hub/commands/jobs.py:
```diff
@@ -0,0 +1,465 @@
+            secrets=self.secrets,
+            flavor=self.flavor,
+            timeout=self.timeout,
+            token=self.token,
```
```suggestion
```

(no need since passed above)

Wauplin reviewed (commented):
I pushed https://github.com/huggingface/huggingface_hub/pull/3211/commits/99b538af3c6129a24ed70770fe2254317306176b to remove `JobUrl` in favor of returning `JobInfo` in all cases (in which I added `self.endpoint`). This makes the API more consistent with other hfh methods.

julien-c commented:
exclude `zero-a10g` from the flavors possibly because i don't think this is a valid hardware here
